Add spec for MinimumOSVersion. (#97)
dotnet/designs: accepted/2020/minimum-os-version/minimum-os-version.md

# .NET 5 OS API Versioning

# Introduction

> The format for [.NET 5 target framework monikers (TFMs)](https://github.com/dotnet/designs/pull/92) includes a version component that represents a revision of the operating system (OS) binding APIs, for example `net5.0-ios15.0` would include a binding to the iOS 15.0 APIs. However, it is a common pattern for apps and libraries to support running on older versions of the OS by using runtime checks to guard their calls so that they do not call APIs that are not available. This applies today for the Xamarin platforms (iOS, Android, Mac etc) and for UWP.

Reviewer: ```suggestion
The format for [.NET 5 target framework monikers (TFMs)](https://github.com/dotnet/designs/blob/master/accepted/2020/net5/net5.md) includes a version component that represents a revision of the operating system (OS) binding APIs. For example, `net5.0-ios15.0` would include a binding to the iOS 15.0 APIs. However, it is a common pattern for apps and libraries to support running on older versions of an OS by using runtime checks to guard their calls so that they do not call APIs that are not available. This applies today for the Xamarin platforms (iOS, Android, Mac etc) and for UWP.
```

> This document proposes standard mechanisms and patterns for runtime API checks and compatibility that can be used across all .NET platforms to increase consistency of developer experience.

Reviewer: ```suggestion
This document proposes a set of standard mechanisms and patterns for runtime API checks and compatibility that can be used across all .NET platforms to increase consistency of developer experience.
```

# OS Versions

There are three different OS version values that may be relevant for apps and libraries: the *OS API Version*, the *OS Target Version*, and the *OS Minimum Version*. It is important to distinguish and clarify these versions, as they have very different purposes.

## OS API Version

> The OS API Version is part of the target framework, and affects which framework reference assemblies the app is compiled against. It does not affect which OS versions the app can run on (for well behaved platforms where backwards ABI compatibility is preserved). The primary reason for an app project to target an old OS API version is to allow the project to continue to build on older versions of the SDK that do not support the more recent API versions. A secondary reason may be to limit the API surface shown in IntelliSense so that it’s easier to avoid newer APIs, though this use case could perhaps be better supported in future by IDE experiences such allowing filtering IntelliSense by platform version.

Reviewer: > The OS API Version is part of the target framework

Would be good to clarify with an example.

> However, for libraries that are distributed as NuGets, it is useful to be able to multi-target across multiple API versions. This allows NuGets to access features available in newer API versions while continuing to ship updates for consumption by apps and libraries that (for whatever reason) target older API versions.

Reviewer: Maybe too niche, but in a big shop, you could also use this same pattern for P2P.

Author: True! I'm not sure it's worth specifically calling it out though

By making the OS API version part of the target framework moniker (TFM), we take advantage of all the existing TFM-based mechanisms for multi-targeting builds against multiple versions of the API, selecting multi-targeted builds of a library from NuGets, and validating binary compatibility of references.

The OS API version has major, minor, servicing and binding revision components. This takes the following form:

```

> {major}[.{minor}[.{servicing}]][-r{bindingRevision}]

Reviewer 2: We got rid of the binding revision component, right?  I think we said these are .NET Framework "Versions" though, ie they have up to 4 version components.

Constants for the platform identifiers could perhaps be exposed on a `PlatformIdentifier` type.

The values declared on this attribute are expected to be accessible by extracting constructor arguments from metadata. However, for readability it’s expected to be applied using subclasses with more specific names. To enforce this, it has an internal constructor and its subclasses are sealed.

The concrete subclasses are `IntroducedAttribute`, `ObsoletedAttribute` and `RemovedAttribute` subclasses, which take of the following form:

```csharp
public sealed class IntroducedAttribute : PlatformAvailabilityAttribute
{
    public IntroducedAttribute(
    string platform,
    int majorVersion,
    int minorVersion = 0,
    int servicingVersion = 0,
    int bindingRevision = 0,
    string message = null
    ) : base(platform, majorVersion, minorVersion, servicingVersion, bindingRevision, message) {}
}
```

These attributes are used as follows:

```csharp
[Introduced(PlatformIdentifier.macOS, 10, 2)]
[Deprecated(PlatformIdentifier.macOS, 10, 9)]

> [Removed(PlatformIdentifier.macOS, 10, 12, 1)]

Reviewer 3: Why not use solutions like `WinMD` and `XLang` to generate metadata (bindings) about the platforms' APIs instead manually duplicating the Swift's attributes into .NET attributes?

That way bindings can be language/framework neutral and in unified format.

Reviewer: This is a super simple and inexpensive mechanism. We don't want to take on heavy weight dependencies for it. It's also "OK" to duplicate attributes like this.

public class Foo : NSObject { ... }
```

The purpose of keeping ‘removed’ APIs and annotating them with `RemovedAttribute` is to allow apps and libraries to continue to use these APIs in fallback paths for older OS versions. This is a pattern that exists on Apple platforms.

If and when C# support target typing for enums, we could add convenience overloads that would improve readability:

```csharp
[Introduced (.macOS, 10, 1)]
[Obsoleted (.macOS, 10, 8)]
[Removed (.macOS, 10, 12)]
```

> NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.

# Runtime Checks

> NOTE: See [this issue](https://github.com/dotnet/runtime/issues/33331) for a more active/current discussion of runtime check APIs.

The following method will be added to the `RuntimeInformation` class in the `System.Runtime.InteropServices` namespace:

```csharp

> public static bool CheckOSVersion(

Author: Yes, it should

> public static bool CheckOSVersion(

Reviewer 4: If we ignore the recommended way to do Windows OS version checks and make this respect the actual OS version, we should fix `Environment.OSVersion` to do the same to be consistent. (I do not have a problem with this - just saying that we should be consistent across the product.)

> public static bool CheckOSVersion(

Reviewer 5: > we should fix Environment.OSVersion to do the same to be consistent. (I do not have a problem with this - just saying that we should be consistent across the product.)

I think we should just do that.  I know it's not the recommended solution, but so many times I've seen folks say that Environment.OSVersion is buggy and then go and just do the P/Invoke themselves (or parse some other string-based representation that's doing it).

> public static bool CheckOSVersion(

Reviewer 4: Opened https://github.com/dotnet/runtime/issues/33643

    int majorVersion, int minorVersion = 0, int servicingVersion = 0
);
```

It is used as follows:

```csharp
if (RuntimeInformation.CheckOSVersion(15)) {

>     // use 15.0+ APIs

Reviewer: Have we thought through the potential for `MissingMethodException`? In the general case, metadata needs to be available to satisfy the JIT. This is not the case for Xamarin.iOS.

Reviewer: I think this is a non-issue, and I'm thinking of challenges from the past. Any platform that supports runtime lightup needs to make sure that it is safe to do so. For example, on Windows, this means that a CSWinRT assembly that targets n Windows version needs to itself expose which APIs are safe to call on n-1. And that's what this spec is describing.

} else {
    // gracefully degrade
}
```

> There will be an additional overload that checks the OS as well as the OS version, for use in portable libraries:

Reviewer: Probably best to define "portable libraries" or use a different term. Is it "libraries that support multiple operating systems via runtime checks?" If yes, do we have a story for providing the compiler with metadata for multiple OSes, for that library?

```csharp
public static bool RuntimeInformation.CheckOS(
    string platformIdentifier,
    int majorVersion, int minorVersion = 0, int servicingVersion = 0
);
```

As with the availability attributes, there could be an additional overload that used enums and target typing to improve readability.

> NOTE: The linker, JIT and/or AOT compiler could treat these method as intrinsics with constant values, which would enable it to eliminate unnecessary fallback code when consuming portable libraries or libraries with a lower OS minimum version than the app itself.

# MSBuild Properties

The `Microsoft.Net.Sdk` targets will extract the OS API Version component from the `TargetFramework` into the `TargetPlatformVersion` MSBuild property and burn it into an assembly attribute. Tools that use this value are expected to access it from the MSBuild property or from the assembly attribute.

> Project files may specify a `MinimumPlatformVersion`, and if they do not it will default to the `TargetPlatformVersion` value. The `MinimumPlatformVersion` must not be higher than the `TargetPlatformVersion`, and this will be validated at the start of the build.

Author: These two specs were written in parallel after we had a discussion a few weeks ago - I don't think either name is canonical right now.

> Project files may specify a `MinimumPlatformVersion`, and if they do not it will default to the `TargetPlatformVersion` value. The `MinimumPlatformVersion` must not be higher than the `TargetPlatformVersion`, and this will be validated at the start of the build.

Reviewer 3: Those property names exist in Windows SDK for a while now. That's why I suggested using those. We can conceptually map those 3 version properties for all platforms, no matter what their platform specific behaviour is.

They differ and that's why we have different platforms 😐. But we don't have to bring that complexity into the user facing project file.

It is recommended that platforms define more specifically named versions of this property of the form `Minimum{PlatformIdentifier}Version`, for example `MinimumIOSVersion` or `MinimumWindowsVersion`, and that their targets assign this value to `MinimumOSVersion` if it is set. This will simplify use of the property multi-targeting.

> NOTE: would `{PlatformIdentifier}MinimumVersion` be better? It would sort/group more easily with other OS-specific properties.

> > NOTE: Using the existing `TargetPlatformVersion` property to represent the API version may be confusing for platforms such as Android that allow having different values for the target version and API version..

Reviewer 6: when would you want to specify a different target and API version?

Reviewer 3: Android always target the API levels. So `*MinVersion` -> becomes minimum API level, `*MaxVersion` becomes targeted API level and we can have `TargetPlatformVersion` point to either of these depending on the project or have new value that mirrors the behaviour of `TargetPlatformMaxVersionTested`.

Before you guys ask, yes, Windows SDK has more than 4 properties just for versions and 3 just for UWP project files!

## Assembly Attribute

The minimum OS version will be burned into the assembly at build time using an assembly attribute. This attribute will be in the `System.Runtime.Versioning` namespace and will take the following form:

```csharp
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple=true)]
public sealed class OSMinimumVersion
{
    public OSMinimumVersion(
    string platformIdentifier,
    int majorVersion, int minorVersion, int servicingVersion
    ) {}
    public string PlatformIdentifier { get; }
    public Version Version { get; }
}
```

As with `PlatformAvailabilityAttribute` , the `PlatformIdentifier` property is included so that platform-agnostic assemblies can participate in platform minimum version annotations and checks.

The MSBuild step that generates and injects TFM assembly attributes will be updated to generate this attribute when the `MinimumPlatformVersion` and `TargetPlatformIdentifier` properties are set.

> Platform-neutral assemblies, i.e. assemblies that target a platform-neutral TFM such as `net5`, will have to declare the attribute explicitly.

Reviewer: ```suggestion
Platform-neutral assemblies, i.e. assemblies that target a platform-neutral TFM such as `net5.0`, will have to declare the attribute explicitly.
```

>
>Perhaps we could use items and metadata instead of platform-specific properties, e.g. `<PlatformInfo Include="ios" MinimumVersion="15.0" />`. For platform-specific projects, we could allow the simpler `MinimumPlatformVersion` property by mapping it into one of these items, e.g. `<PlatformInfo Include="$(TargetPlatformIdentifier)" Version="$(MinimumPlatformVersion)" Condition="'$(TargetPlatformIdentifier)' != ''" />`.

## NuGet Manifest

When generating a NuGet package using MSBuild, any platform minimum version constraints expressed in the project file will be listed in the `.nuspec`  manifest in a `<platformInfo>` child of the `<metadata>` element. This element and all of its children are optional. They expose the same information as the assembly annotation attributes, but avoid the need for consumers to read the assembly metadata. This could potentially be validated and/or automated by NuGet..

> Within the `<platformInfo>` there are child elements corresponding to the TFM short names (used by NuGet as the folder name) of the reference assemblies in the NuGet. If the element represents a platform-specific TFM (e.g. `net5-ios14.0` or `net6-android9.0`) then it must have a `minimumVersion` attribute representing the minimum platform version for that platform:

Reviewer: ```suggestion
Within the `<platformInfo>` element, there are child elements corresponding to the TFM short names (used by NuGet as the folder name) of the reference assemblies in the NuGet. If the element represents a platform-specific TFM (e.g. `net5-ios14.0` or `net6-android9.0`) then it must have a `minimumVersion` attribute representing the minimum platform version for that platform:
```

```xml
<metadata>
    <platformInfo>

>     <net5-ios15.0 minimumVersion="13.0" />
>     <net5-mac10.15 minimumVersion="10.11" />

Reviewer: ```suggestion
    <net5.0-ios15.0 minimumVersion="13.0" />
    <net5.0-mac10.15 minimumVersion="10.11" />
```

> <metadata>
>     <platformInfo>
>     <net5-ios15.0 minimumVersion="13.0" />
>     <net5-mac10.15 minimumVersion="10.11" />
>     </platformInfo>
> </metadata>

Reviewer 3: ```suggestion
<metadata>
    <platforms framework="net5.0">
      <platform id="ios" minVersion="13.0" targetVersion="13.2" maxVersion="14.0">
      <platform id="mac" minVersion="10.10" targetVersion="10.11" maxVersion="10.14">
      <platform id="android" minVersion="18" targetVersion="21" maxVersion="24">
    </platforms>
    <platforms framework="net6.0">
      <platform id="ios" minVersion="13.1" targetVersion="13.3" maxVersion="14.0">
      <platform id="mac" minVersion="10.11" targetVersion="10.12" maxVersion="10.14">
      <platform id="android" minVersion="19" targetVersion="22" maxVersion="24">
    </platforms>
</metadata>
```

We can use this format for the platform neutral declarations too (see **below**). These are only needed if we have interop between native lib and a managed implementation but could be useful for binding libs too as they are essentially more or less the same!

`targetVersion` and `maxVersion` are totally optional. They could used to further manage the dependencies and resolve conflicts for large typical project that targets multiple platforms.

```

> A reference assembly for a platform-neutral TFM (e.g. `net5` or `net6`) may still need to express platform-specific minimum versions if the implementations are platform-specific (aka ‘bait and switch’). As before, the element matches the TFM of the reference assembly, but this time there is no `minimumVersion`. Instead there are sub-elements that provide `minimumVersion` values for platforms using the `TargetPlatformIdentifier` of those platforms to identify them:

Reviewer: ```suggestion
A reference assembly for a platform-neutral TFM (e.g. `net5.0` or `net6.0`) may still need to express platform-specific minimum versions if the implementations are platform-specific (aka ‘bait and switch’). As before, the element matches the TFM of the reference assembly, but this time there is no `minimumVersion`. Instead there are sub-elements that provide `minimumVersion` values for platforms using the `TargetPlatformIdentifier` of those platforms to identify them:
```

Reviewer: Honest question ... do we have satisfactory support for bait-and-switch in dotnet pack today? If not, it should noted.

Author: No, we do not. That was one of the things that NuGetizer did well.

```xml
<metadata>
    <platformInfo>
    <net6>
        <ios minimumVersion="13.0">
        <mac minimumVersion="11.0">
    </net6>
    </platformInfo>
</metadata>
```

> Using these mechanisms, a NuGet package can declare minimum platform versions for any or all of its reference assemblies.

Reviewer 3: Why do we want to declare target platform for neutral assemblies? They are platform and runtime independent, no?

Reviewer: This is covered a bit earlier in the doc. The vision is that neutral assemblies can still do guarded light-up across a set of OSes.

## OS Minimum Version

The OS Minimum Version is the minimum version of the OS that an app or library supports running on. If it is not specified, it should default to the OS API Version. If an app or library author specifies an OS Minimum Version, they are declaring that they are taking responsibility for ensuring that the app can run on that version and higher by using runtime checks to guard calls to potentially unavailable APIs. This allows apps and libraries to progressively “light up” features on newer OS versions while continuing to run downlevel.

We will make this concept first class and propagate the value throughout the build system and NuGet consistently, which will enable us to build analyzers and other experiences to help developers follow this pattern. This tooling will be detailed later in this document.

> # API Annotations

Reviewer 7: It might be helpful to outline our guidelines for 3rd party TFM libraries. 
- should they always have the annotation
- or only when they directly call to platform api
- or only when they wrap annotated api
- etc

Author: Added some very basic guidance on this

To enable tooling to provide a good experience, the reference assemblies must be annotated with information about the availability of the APIs, as is current done for Xamarin Apple platforms.

The following attributes are proposed for the `System.Runtime.Versioning` namespace. They are based on the Xamarin attributes, but they do not have the `Deprecated` availability kind (it’s redundant), they rename the `Unavailable` availability kind to `Removed` (for consistency), and they do not have the ability to differentiate availability based on architecture (reduction of scope).

```csharp
[AttributeUsage(AttributeTargets.All, AllowMultiple=true)]
public abstract class PlatformAvailabilityAttribute
{
    internal PlatformAvailabilityAttribute(
    string platformIdentifier,
    int majorVersion, int minorVersion, int servicingVersion, int bindingRevision
    string message
    ) {}
    public AvailabilityKind AvailabilityKind { get; }
    public string Message { get; }
    public string PlatformIdentifier { get; }
    public Version OSVersion { get; }
    public int BindingRevision { get; }
}

public enum AvailabilityKind
{
    Introduced,
    Obsoleted,
    Removed
}
```

> The platform string is expected to have the same values as used in the `TargetPlatformIdentifier` portion of platform-specific TFMs. Having it in the availability attributes may seem redundant as it will also be burned into assembly attributes for platform-specific reference assemblies. However, explicitly specifying the attribute is useful for platform-agnostic `net5.0` reference assemblies that have methods that are only available on some platforms e.g. due to having platform-specific implementations via bait-and-switch. This may also improve source readability when multi-targeting as it will remove the need to use `#if` around the annotations.

Reviewer 7: @terrajobst is the thinking to apply such attributes to core libraries as well? For example when OpenSSL requires specific underlying version, or security APIs relying on specific PAL ios/macOS version?

Reviewer 3: Even for platform-agnostic assemblies using platform-specific implementations should match the public surface APIs (aka the ref assembly should be one and only one for any number of platform specific implementations!)

But that need not be the case for platform-specific assemblies that might have different public APIs for each platform.

Reviewer: Yes, I would assume we would add these to `net5.0` TFM.

Constants for the platform identifiers could perhaps be exposed on a `PlatformIdentifier` type.

The values declared on this attribute are expected to be accessible by extracting constructor arguments from metadata. However, for readability it’s expected to be applied using subclasses with more specific names. To enforce this, it has an internal constructor and its subclasses are sealed.

The concrete subclasses are `IntroducedAttribute`, `ObsoletedAttribute` and `RemovedAttribute` subclasses, which take of the following form:

```csharp

> public sealed class IntroducedAttribute : AvailabilityAttribute

Reviewer 8: The base class is `PlatformAvailabilityAttribute` from above: 

```suggestion
public sealed class IntroducedAttribute : PlatformAvailabilityAttribute
```

{
    public IntroducedAttribute(
    string platform,
    int majorVersion,
    int minorVersion = 0,
    int servicingVersion = 0,
    int bindingRevision = 0,
    string message = null
    ) : base(platform, majorVersion, minorVersion, servicingVersion, bindingRevision, message) {}
}
```

These attributes are used as follows:

```csharp
[Introduced(PlatformIdentifier.macOS, 10, 2)]
[Deprecated(PlatformIdentifier.macOS, 10, 9)]
[Removed(PlatformIdentifier.macOS, 10, 12, 1)]
public class Foo : NSObject { ... }
```

The purpose of keeping ‘removed’ APIs and annotating them with `RemovedAttribute` is to allow apps and libraries to continue to use these APIs in fallback paths for older OS versions. This is a pattern that exists on Apple platforms.

If and when C# support target typing for enums, we could add convenience overloads that would improve readability:

```csharp
[Introduced (.macOS, 10, 1)]
[Obsoleted (.macOS, 10, 8)]
[Removed (.macOS, 10, 12)]
```

> > NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.

Reviewer 7: I don't think we even need to do that. The attributes will exist in reference assemblies only which are not processed by illinker.

> > NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.

Reviewer 9: For the BCL itself it has not impact, but for platform assemblies (our main use case) we need the attributes in the implementation assemblies since they are heavily used by our test suite.

> > NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.

Reviewer 7: I think that's a very specific use case and size increase does not matter in this case. I also don't think everyone should pay the penalty of having the attributes in implementation libs just because it makes your testing simpler.

> > NOTE: These attributes would only be needed on reference assemblies. They would not be needed at runtime. They could be removed by the linker, and we could also strip them from the implementation assemblies that we ship ourselves.

Author: Added a note, but I'm not sure how we could make it easy to omit these from reference assemblies, though.

> NOTE: The linker, JIT and/or AOT compiler could treat these method as intrinsics with constant values, which would enable it to eliminate unnecessary fallback code when consuming portable libraries or libraries with a lower OS minimum version than the app itself.

# MSBuild Properties

The `Microsoft.Net.Sdk` targets will extract the OS API Version component from the `TargetFramework` into the `TargetPlatformVersion` MSBuild property and burn it into an assembly attribute. Tools that use this value are expected to access it from the MSBuild property or from the assembly attribute.

Project files may specify a `MinimumPlatformVersion`, and if they do not it will default to the `TargetPlatformVersion` value. The `MinimumPlatformVersion` must not be higher than the `TargetPlatformVersion`, and this will be validated at the start of the build.

It is recommended that platforms define more specifically named versions of this property of the form `Minimum{PlatformIdentifier}Version`, for example `MinimumIOSVersion` or `MinimumWindowsVersion`, and that their targets assign this value to `MinimumOSVersion` if it is set. This will simplify use of the property multi-targeting.

> > NOTE: would `{PlatformIdentifier}MinimumVersion` be better? It would sort/group more easily with other OS-specific properties.

Reviewer 3: When doing multi-targeting, we can condition the `TargetPlatform*Version` properties under `TargetFramework` (aka multi-targeting) property (still don't like the name, BTW, so confusing!). But for single-targeting projects, we don't need to because, you know, they only target one thing.

> NOTE: Using the existing `TargetPlatformVersion` property to represent the API version may be confusing for platforms such as Android that allow having different values for the target version and API version..

# Assembly Level Minimum Version

We cannot provide useful errors or warnings at an assembly level when a project references a library with a higher minimum version, as it may guard usage of that library with runtime checks. 

However, it still useful to burn the minimum platform version into the assembly or NuGet package as a a baseline value for the classes and members that are not explicitly annotated. It can also be displayed as an information value in the user experience, for example in the NuGet package manager or on NuGet.org, although those are out of the scope of this document.

## Assembly Attribute

> The minimum OS version will be burned into the assembly at build time using an assembly attribute. This attribute will be in the `System.Runtime.Versioning` namespace and will take the following form:

Reviewer 3: Why not let the Platform runtime (if there will be) take care of roll-forwarding just like the coreclr does now?

It'd be way simpler just to define `TargetPlatform` as analogues to `TargetFramework` attribute which is already there and let the `runtimeconfig` take care of the roll-forwarding.

> The minimum OS version will be burned into the assembly at build time using an assembly attribute. This attribute will be in the `System.Runtime.Versioning` namespace and will take the following form:

Author: It's only intended for compile time usage

```csharp
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple=true)]

> public sealed class OSMinimumVersion

Author: Not sure tbh? It's not as closely related

```xml
<metadata>
    <platformInfo>
    <net5-ios15.0 minimumVersion="13.0" />
    <net5-mac10.15 minimumVersion="10.11" />
    </platformInfo>
</metadata>
```

A reference assembly for a platform-neutral TFM (e.g. `net5` or `net6`) may still need to express platform-specific minimum versions if the implementations are platform-specific (aka ‘bait and switch’). As before, the element matches the TFM of the reference assembly, but this time there is no `minimumVersion`. Instead there are sub-elements that provide `minimumVersion` values for platforms using the `TargetPlatformIdentifier` of those platforms to identify them:

```xml
<metadata>
    <platformInfo>
    <net6>
        <ios minimumVersion="13.0">
        <mac minimumVersion="11.0">
    </net6>
    </platformInfo>
</metadata>
```

Using these mechanisms, a NuGet package can declare minimum platform versions for any or all of its reference assemblies.

# Availability Inheritance

> Members and nested types that do not have explicit availability annotations will inherit the values from their containing types. Types that do not have `Introduced` annotations will be treated as if they were introduced in the minimum platform version of the assembly itself.

Reviewer 3: Can't we have separate assembly for Legacy APIs. I really like the concept of jQuery and jQuery Migrate which allows you to slowly migrate away from obsolete APIs. Sure, Analysers are good and all but when considering IoT and WASM targets, size and performance is everything.

So for E.g., if we can have APIs like string and string format overloads moved to legacy libs in favour of string interpolation APIs, that would reduce the bloat of the libs by a significant factor!!

> Please @ or # link the relevant issues if you can find them.

> Members and nested types that do not have explicit availability annotations will inherit the values from their containing types. Types that do not have `Introduced` annotations will be treated as if they were introduced in the minimum platform version of the assembly itself.

Author: @Nirmal4G that's somewhat orthogonal to this proposal, but check out https://github.com/dotnet/designs/blob/master/accepted/2020/better-obsoletion/better-obsoletion.md

> Members and nested types that do not have explicit availability annotations will inherit the values from their containing types. Types that do not have `Introduced` annotations will be treated as if they were introduced in the minimum platform version of the assembly itself.

Reviewer 3: Not the solution I was hoping for for but 😐!

# IDE Experience

Quick Info and IntelliSense tooltips will display the minimum platform version for any member with a minimum platform version higher than the current project’s minimum platform version.

We may also want to consider some kind of sorting, grouping or colorization based hints in the IntelliSense list.

# Analyzers

There will be a Roslyn analyzer that will squiggle calls to any method or property with a minimum platform version higher than the current project’s minimum platform version, unless either of the following is true:

- the call is guarded by a `CheckOS` or `CheckOSVersion` call against a version that is greater than or equal to the member’s minimum platform version

> - the caller member has a minimum platform version that is equal to or higher than the callee

Reviewer 9: There's potential to have analyzers to ensure availability attributes are logical - but it brings some special cases.

E.g. when a new type is introduced in a hierarchy, which is not a breaking change, you can end up with

```csharp
[Introduced (.ios, 8,0)]
class NSFoo : NSObject {
   string Id {get;}
   string Name {get;}
}
```
becoming
```csharp
[Introduced (.ios, 14,0)]
class NSMiddle : NSObject {
   [Introduced (.ios, 8,0)]
   string Id {get;}
}

[Introduced (.ios, 8,0)]
class NSFoo : NSMiddle {
   string Name {get;}
}
```

and it's easy to think there's a mistake in `Middle` since a member was available well before it's type. While you can't create `NSMiddle` on iOS8 you can still use `NSFoo.Middle` (without warnings) on iOS8.

Reviewer 3: We can use [Windows](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/version-adaptive-code)'s existing version and api tagging or we could go with existing tags which are already there to expose them as such in a more .NET/WinMD way. Existing tools could be easily supported if we did follow latter idea.

> Try to use ideas as much as from the Windows platform teams. They are relevant to all other platforms out there.

}
```

There will be another method on `RuntimeInformation` that checks the OS as well as the OS version, for use in platform-neutral libraries that use runtime checks to support or light up features on multiple OSes:

```csharp
public static bool CheckOS(
    string platformIdentifier,
    int majorVersion, int minorVersion = 0, int servicingVersion = 0
);
```

As with the availability attributes, there could be an additional overload that used enums and target typing to improve readability.

***NOTE****: The linker, JIT and/or AOT compiler could treat these method as intrinsics with constant values, which would enable it to eliminate unnecessary fallback code when consuming portable libraries or libraries with a lower OS minimum version than the app itself.*

# MSBuild Properties

The `Microsoft.Net.Sdk` targets will extract the OS API Version component from the `TargetFramework` into the `TargetPlatformVersion` MSBuild property and burn it into an assembly attribute. Tools that use this value are expected to access it from the MSBuild property or from the assembly attribute.

> Project files may specify an `OSMinimumVersion`, and if they do not it will default to OS Version equivalent to the OS API version specified in the `TargetPlatformVersion` value. The `OSMinimumVersion` must not be higher than the `TargetPlatformVersion`, and this will be validated at the start of the build.

Reviewer 2: How will we handle this if we update the bindings outside of the OS version.  IE we ship iOS bindings 15.0.1, but it still maps to iOS 15.0.  Does each workload need to keep a mapping between `TargetPlatformVersion` and `OSMinimumVersion` for that platform?

Author: Yes, that's unfortunate but necessary. Still thinking about whether to formally define a standard mechanism or leave it up to the platform targets.

Reviewer 3: Why not define a roll forward mechanism for bindings just like the runtime packages?

That would certainly remove these complications!!

Author: @Nirmal4G it's not that easy.

For example: from the TFM `net5-ios15.0` we can infer that the `OSMinimumVersion` default is `15.0`. However, suppose the TFM `net5-ios15.0.1` was a revision of the bindings rather than bindings for a new OS version, then its `OSMinimumVersion` would also be need to be `15.0`.

We could handle that in a general way by saying that only the first two components of the binding version are the OS version and the third component is the bindings revision. However, those kinds of conventions we need to be determined on a platform by platform basis, as different platforms have different versioning schemes.

Reviewer 3: This is still a problem in the current proposal with the TF having bindings revision. Do we expect people to target the same platform for each bindings revision, I don't think so. I didn't encounter any packages targeting specifically any platform with multiple bindings!!! Are there?

If so, Why not separate the bindings version into a separate property?

Say `PlatformPackageVersion` if we're using nuget as the delivery mechanism for the bindings and `PlatformBindingsRevision` if we're shipping bindings within the workload itself. If we're using both we could use the latter to set the former for resolving the implicit FrameworkReference.

It is recommended that platforms define more specifically named versions of this property of the form `{PlatformIdentifier}MinimumVersion`, for example `IOSMinimumVersion` or `WindowsMinimumVersion`, and that their targets assign this value to `OSMinimumVersion` if it is set. This will simplify use of the property when multi-targeting.

These per-platform properties will not work when building platform-neutral assemblies, as that would require centralizing definitions of all of the platform-specific `MinimumVersion` properties into the `Microsoft.Net.Sdk` targets. We may at some point add some other way to specify `OSMinimumVersion` values for multiple OSes for a single platform-neutral assembly, perhaps using items, but for now such assemblies must use assembly attributes directly.

> > ***NOTE***: `{PlatformIdentifier}MinimumVersion` was chosen over `Minimum{PlatformIdentifier}Version` as it it's more easily searchable and sorts/group more easily with other OS-specific properties.

Reviewer 2: I think Windows already has a property defined that follows a different format: `TargetPlatformMinVersion`.  We should cover how that relates to the new properties.

Author: Added a clarifying note - https://github.com/dotnet/designs/pull/97/commits/7c34d18720ec6c25374321ee8bb2a21e74eae99a /cc @terrajobst

Reviewer 3: Apple platforms and Windows defines the same version for both API and OS. But Android has API levels. But do the MSBuild projects (and the Android SDK tools) need the OS version? If so, Why?

As with `PlatformAvailabilityAttribute` , the `PlatformIdentifier` property is included so that platform-agnostic assemblies can participate in platform minimum version annotations and checks.

The MSBuild step that generates and injects TFM assembly attributes will be updated to generate this attribute when the `OSMinimumVersion` and `TargetPlatformIdentifier` properties are set.

Platform-neutral assemblies, i.e. assemblies that target a platform-neutral TFM such as `net5.0`, will have to declare the attribute explicitly.

Like the API annotation attributes, the `OSMinimumVersionAttribute` attribute is not needed in implementation assemblies as it is not used at runtime.

## NuGet Manifest

> When generating a NuGet package using MSBuild, any platform minimum version constraints expressed in the project file will be listed in the `.nuspec` manifest in a `<platformInfo>` child of the `<metadata>` element. This element and all of its children are optional. They expose the same information as the assembly annotation attributes, but avoid the need for consumers to read the assembly metadata. This could potentially be validated and/or automated by NuGet..

Reviewer 2: Nit: extra period at end of paragraph.
